src/storage/UpdateEdgeProcessor.cpp

```diff
@@ -0,0 +1,373 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "storage/UpdateEdgeProcessor.h"
+#include "base/NebulaKeyUtils.h"
+#include "dataman/RowWriter.h"
+#include "kvstore/LogEncoder.h"
+
+namespace nebula {
+namespace storage {
+
+void UpdateEdgeProcessor::onProcessFinished(int32_t retNum) {
+    if (retNum > 0) {
+        nebula::cpp2::Schema respScheam;
+        respScheam.columns.resize(retNum);
+        for (auto& prop : this->edgeContext_.props_) {
+            if (prop.returned_) {
+                auto column = this->columnDef(std::move(prop.prop_.name), prop.type_.type);
+                respScheam.columns[prop.retIndex_] = std::move(column);
+            }
+        }
+        for (auto& tc : this->tagContexts_) {
+            for (auto& prop : tc.props_) {
+                if (prop.returned_) {
+                    auto column = this->columnDef(std::move(prop.prop_.name), prop.type_.type);
+                    respScheam.columns[prop.retIndex_] = std::move(column);
+                }
+            }
+        }
+        resp_.set_schema(std::move(respScheam));
+        resp_.set_data(respData_);
+    }
+}
+
+
+kvstore::ResultCode UpdateEdgeProcessor::collectEdgesProps(
+                            PartitionID partId,
+                            const cpp2::EdgeKey& edgeKey,
+                            const std::vector<PropContext>& props,
+                            FilterContext* fcontext,
+                            Collector* collector) {
+    auto prefix = NebulaKeyUtils::prefix(partId, edgeKey.src, edgeKey.edge_type,
+                                         edgeKey.ranking, edgeKey.dst);
+    std::unique_ptr<kvstore::KVIterator> iter;
+    auto ret = kvstore_->prefix(spaceId_, partId, prefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        VLOG(3) << "Error! ret = " << static_cast<int32_t>(ret) << ", spaceId " << spaceId_;
+        return ret;
+    }
+    // Only use the latest version.
+    if (iter && iter->valid()) {
+        key_ = iter->key().toString();
+        edgeReader_ = RowReader::getEdgePropReader(schemaMan_,
+                                                   iter->val(),
+                                                   spaceId_,
+                                                   edgeKey.edge_type);
+        filterReader_ = RowReader::getEdgePropReader(schemaMan_,
```

Comment:
Why we have two readers?

Reply:
The edgeReader_ is used to transfer to updater which is std::move. The filterReader_ is used to get the value of Expression.

Reply:
Could we unify them

Reply:
I have no idea to unify them, do you have any advice?

Reply:
It has been refactored.